File: macos/Services/MCP/Tools/Tier1/ListConnectionsTool.swift
// ListConnectionsTool.swift
// Gridex
//
// MCP Tool: List all configured database connections.

import Foundation

struct ListConnectionsTool: MCPTool {
    let name = "list_connections"
    let description = "List all configured database connections. Returns connection IDs, names, and types (postgres/mysql/sqlite/redis/mongodb/mssql)."
    let tier = MCPPermissionTier.schema

    let inputSchema: [String: Any] = [
        "type": "object",
        "properties": [:] as [String: Any]
    ]

    func execute(params: JSONValue, context: MCPToolContext) async throws -> MCPToolResult {
        let configs = try await context.connectionRepository.fetchAll()
        let activeConnections = await context.connectionManager.allActiveConnections

        var connections: [[String: Any]] = []
        for config in configs {
            let mode = await context.permissionEngine.getMode(for: config.id)

            // Skip locked connections from MCP view
            if mode == .locked { continue }

            let isConnected = activeConnections.contains { $0.id == config.id }

            connections.append([
                "id": config.id.uuidString,
                "name": config.name,
                "type": config.databaseType.rawValue,
                "host": config.displayHost,
                "database": config.database ?? "",
                "is_connected": isConnected,
                "mcp_mode": mode.rawValue
            ])
        }

        if connections.isEmpty {
            return MCPToolResult(text: "No connections available for MCP access. All connections are either locked or none exist.")
        }

        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        let data = try encoder.encode(connections.map { dict -> [String: String] in
            dict.mapValues { "\($0)" }
        })
        let json = String(data: data, encoding: .utf8) ?? "[]"

        return MCPToolResult(text: json)
    }
}

File: windows/Gridex/Services/MCP/Tools/Tier1/ListConnectionsTool.cpp
//
// ListConnectionsTool.cpp
//

#include "ListConnectionsTool.h"
#include "../MCPToolHelpers.h"
#include "../../../../Models/ConnectionStore.h"

namespace DBModels
{
    MCPToolResult ListConnectionsTool::execute(const nlohmann::json&, MCPToolContext& ctx)
    {
        const auto configs = ConnectionStore::Load();
        nlohmann::json arr = nlohmann::json::array();

        for (const auto& c : configs)
        {
            // Hide Locked rows entirely — mac parity.
            const auto mode = ctx.permissionEngine.getMode(c.id);
            if (mode == MCPConnectionMode::Locked) continue;

            nlohmann::json entry = {
                {"id",        MCPToolHelpers::toUtf8(c.id)},
                {"name",      MCPToolHelpers::toUtf8(c.name)},
                {"type",      MCPToolHelpers::toUtf8(DatabaseTypeDisplayName(c.databaseType))},
                {"host",      MCPToolHelpers::toUtf8(c.host)},
                {"database",  MCPToolHelpers::toUtf8(c.database)},
                {"mcp_mode",  mcpRawString(mode)}
            };
            arr.push_back(std::move(entry));
        }

        if (arr.empty())
            return MCPToolResult::text(
                "No connections available for MCP access. All connections are "
                "either locked or none exist.");

        return MCPToolResult::text(arr.dump(2));
    }
}
